### PR TITLE
Attention is all you need

### PR DESCRIPTION
**Pull request by ivanleomk**

### CODE REVIEW

File: code/attention-is-all-you-need/attention.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    def __call__(self, x: np.array, mask=False) -> np.array:
+        """
+        This is how we can call our forward pass function
+        """
+        assert (
+            x.shape[1] == self.d_input
+        ), f"Passed in an array of size {len(x)}. Maximum context size is {self.d_input}"
+        q_prime = x @ self.w_q.T  # x : (d_input , d_k ) , w_q.T : (d_k , d_input )
+        k_prime = x @ self.w_k.T  # x : (d_input , d_k ) , w_k.T : (d_k , d_input )
+        v_prime = x @ self.w_v.T
```
Comment (paulinamoskwa) [inline code/attention-is-all-you-need/attention.py]: little doubt: isn't this part of the multi-head attention? 🤔

(as of my understanding) scaled dot product attention just takes {Q, K, V} and operates on them as they are. From the paper I understood that the projection onto other spaces (so the multiplication with {W_Q, W_K, W_V}) happens before the self-attention is called 

image

Comment (ivanleomk) [reply code/attention-is-all-you-need/attention.py]: Hey @paulinamoskwa, on closer look, yes you're right, I confused Self-Attention and Scaled Dot Product Attention. 

Scaled dot product attention just takes {Q, K, V} and operates on them as they are. Self-Attention on the other hand takes in an input vector x with dim ( n, d_embeddings ) and then outputs a final output embedding of dim (n, d_embddings) with the help of the linear layer which does a transformation from the output of the Scaled Dot Product Attention 

Seems like my code here implements a Multi-Head attention block with a single step ( without the linear layer ) instead. I wanted to follow Raschka's implementation of Self-Attention from this [Video](https://www.youtube.com/watch?v=0PjHri8tc1c) which has this specific graphic

CleanShot 2024-01-21 at 08 38 57

Thanks for the feedback!

File: pages/papers/attention-is-all-you-need.mdx [hunk 2]
```diff
@@ -1,3 +1,223 @@
...
+
+It's prohibitively expensive to store an equivalent one-hot encoded vector. Each row would have 37,000 values, most of which would just be 0. 
+
+Secondly, embeddings can capture more information due to their real-number values, making them an ideal choice.
+
+**In our case, each token has an embedding dimension of 512, or 512 floats for each individual token.**
+
+### Positional Encoding
+
+> The original transformer used a sinusodial embedding which gives a more absolute concept of position. Current work with implementations such as Rotatory Embeddings provide a way for transformers to learn more relative representations of positions in the sequence.
```
Comment (ivanleomk) [inline pages/papers/attention-is-all-you-need.mdx:80]: @aimuggle check if this is ok? Wanted to bring in your point here

File: pages/papers/attention-is-all-you-need.mdx [hunk 3]
```diff
@@ -1,3 +1,223 @@
...
+$$
+
+Note that Q,K and V do not have the be the same matrix. Q and K will have the same dimensions `d_k` and V will have the dimensions of `d_v` . In the paper, the model has an embedding dimension `d_model` of `512` where `d_k = d_v = d_model` .
+
+There are two important things to note here
+
+1. We scale $QK^T$ by $\sqrt{d_k}$ because of numerical instability. This is because the product of two matrices has a significantly greater variance and tends to lead to large values being generated. Scaling down allows for a smaller value to be used when computing the softmax, resulting in better results
+2. The Dot Product is chosen because it is much faster and more space-efficient in practice, since it can be implemented using highly optimised matrix multiplication code.
+
+The model tunes the values of Q and K so that it learns how to assign the proper weightage to each token for a given sequence. We can intuitively understand the Query matrix as a representation of a sequence seeking a match while the Key matrix is a representation of a sequence's meaning.
```
Comment (ivanleomk) [inline pages/papers/attention-is-all-you-need.mdx:121]: Not too sure how to phrase the "seeker" and the has a value representation

File: pages/papers/attention-is-all-you-need.mdx [hunk 4]
```diff
@@ -1,3 +1,223 @@
...
+The first is Self-Attention, where the Q, K and V matrices are derived from the same input vector (As seen above). However, we also have Cross-Attention, where the Q,K and V matrices are not. 
+
+In the example above, our decoder Multi-Head attention block derives Q and K from the encoder’s output and V from the output of the previous Multi-Head Attention block. 
+
+> Note that the same encoder output is passed to every single decoder block. Hence, information is encoded once and then cached for every subsequent inference or training call
+>
+
+## Encoders and Decoders
+
+> Most transformer models nowadays are decoder only because decoders tend to shine more at text generation work. Encoders helps to bridge the gap between sequences that exist in two separate embedding spaces - which is where encoder-decoder architectures tend to shine.
```
Comment (ivanleomk) [inline pages/papers/attention-is-all-you-need.mdx:158]: Bringing in some additional stuff here

File: pages/papers/attention-is-all-you-need.mdx [hunk 5]
```diff
@@ -1,3 +1,223 @@
...
+
+## Encoders and Decoders
+
+> Most transformer models nowadays are decoder only because decoders tend to shine more at text generation work. Encoders helps to bridge the gap between sequences that exist in two separate embedding spaces - which is where encoder-decoder architectures tend to shine. 
+
+![Transformer Architecture](/transformer-architecture.png)
+
+An encoder maps an input sequence of symbolic representations to a sequence of continous representations. This is then used by the decoder to generate an output sequence in an auto-regressive manner by consuming the previously generated symbols as additional input when generating the next token.
+
+Using an Encoder and Decoder which separately parse an input and output sequence allows our model to work with two fundamentally different embedding spaces. This is because the **Output and Input Embeddings can be totally different.** This results in different vocabulary sizes and different representations. This allows our model to better learn the features of each language and in turn be able to closely relate them
```
Comment (ivanleomk) [inline pages/papers/attention-is-all-you-need.mdx:164]: Was thinking of what bryan was saying about mapping between languages that have diff structural properties

eg. subject - verb -agreement vs subject agreement verb or something like that

File: pages/papers/attention-is-all-you-need.mdx [hunk 6]
```diff
@@ -1,3 +1,223 @@
...
+These are the notes which I took on the paper - Attention is all you need. PRs are welcome if I've made any errors. 
+
+## Motivation
+
+> Another intuitive way we can think of transformers are as classifiers where they determine what is the most likely next token
+
+Transformers were introduced in 2017 in Attention Is All You Need and provided a strong alternative to RNNs. There are three main advantages of using transformers as compared to RNNs
+
+1. **Parallelisation**: RNNs are sequential in nature, therefore the runtime of your training/inference scales linearly with the length of your sequence. This makes it difficult to effectively batch training/inference calls 
+2. **Performance** : Because RNNs process information by sequentially updating a hidden state, this makes it difficult for the model to quick access past and future state. Lookup is `O(n)` by definition
```
Comment (blackbeelabs) [inline pages/papers/attention-is-all-you-need.mdx:16]: NIT: quick -> quickly

File: pages/papers/attention-is-all-you-need.mdx (hunk 2 above)
Comment (blackbeelabs) [reply pages/papers/attention-is-all-you-need.mdx:80]: recommend to see https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1224/slides/cs224n-2022-lecture09-transformers.pdf slides 40/41, and https://web.stanford.edu/class/cs224n/readings/cs224n-self-attention-transformers-2023_draft.pdf pg 6 for details

File: pages/papers/attention-is-all-you-need.mdx [hunk 7]
```diff
@@ -1,3 +1,223 @@
...
+
+1. There was a car in front of me, it was red
+2. There was a car in front of me, it was challenging to swerve around it
+
+There are a few things to notice about this sentence
+
+- `it` appears twice - but it refers to different things each time. In the first instance, it refers to the act of driving around the car while the second instance refers to the car.
+- `it` is heavily context dependent, in the first sentence, `it` refers to the car instead!
+
+As a result, we use positional encodings to imbue our model's embeddings with information on the position of each word. The model learns these transformations over time. These encodings are computed once and then frozen in time.
```
Comment (blackbeelabs) [inline pages/papers/attention-is-all-you-need.mdx:94]: nit: ...position of each word relative to the input sequence

File: pages/papers/attention-is-all-you-need.mdx [hunk 8]
```diff
@@ -1,3 +1,223 @@
...
+
+$$
+Attention(Q,K,V) = softmax(\frac{QK^{T}}{\sqrt{d_k}})V
+$$
+
+Note that Q,K and V do not have the be the same matrix. Q and K will have the same dimensions `d_k` and V will have the dimensions of `d_v` . In the paper, the model has an embedding dimension `d_model` of `512` where `d_k = d_v = d_model` .
+
+There are two important things to note here
+
+1. We scale $QK^T$ by $\sqrt{d_k}$ because of numerical instability. This is because the product of two matrices has a significantly greater variance and tends to lead to large values being generated. Scaling down allows for a smaller value to be used when computing the softmax, resulting in better results
```
Comment (blackbeelabs) [inline pages/papers/attention-is-all-you-need.mdx:118]: (consider to add) Smaller values generally have less variance, and are hence numerically more stable.

File: pages/papers/attention-is-all-you-need.mdx (hunk 4 above)
Comment (blackbeelabs) [reply pages/papers/attention-is-all-you-need.mdx:158]: consider to add https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1224/slides/cs224n-2022-lecture10-pretraining.pdf ref. slide 28 on examples

File: pages/papers/attention-is-all-you-need.mdx (hunk 3 above)
Comment (blackbeelabs) [reply pages/papers/attention-is-all-you-need.mdx:121]: consider pg 4 of https://web.stanford.edu/class/cs224n/readings/cs224n-self-attention-transformers-2023_draft.pdf
and slide 28 of https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1224/slides/cs224n-2022-lecture09-transformers.pdf

In order to build a contextual representation of a token in the sequence, the model finds out what information in the values (V) are relevant. It does so by comparing the Q against all K and from this comparison, finds the relative importance of each V to build the contextual state.
